### PR TITLE
Makes ResourceSurrogateId derived from the modification timestamp

### PR DESCRIPTION
**Pull request by johnstairs**
## Description
The surrogate ID we use for resources will now be derived from the modification timestamp. This allows predicates over `_lastUpdated` to be served using the `ResourceSurrogateId` column, which is already present and indexed across all tables.

With this change, the last updated timestamp given back to the client will always be truncated to millisecond precision. We then use a cycling sequence to fill the lower bits to make the ID unique

## Related issues
Addresses [AB#69345](https://microsofthealth.visualstudio.com/f8da5110-49b1-4e9f-9022-2f58b6124ff9/_workitems/edit/69345)

## Testing
Describe how this change was tested.

### CODE REVIEW

File: src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/LastUpdatedToResourceSurrogateIdRewriter.cs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+using System;
+using Microsoft.Health.Fhir.Core.Extensions;
+using Microsoft.Health.Fhir.Core.Features.Search;
+using Microsoft.Health.Fhir.Core.Features.Search.Expressions;
+using Microsoft.Health.Fhir.SqlServer.Features.Storage;
+
+namespace Microsoft.Health.Fhir.SqlServer.Features.Search.Expressions.Visitors
+{
+    /// <summary>
+    /// Turns predicates over _ka
```
Comment (cunninghamjc) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/LastUpdatedToResourceSurrogateIdRewriter.cs]: _ka?

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/1.sql [hunk 2]
```diff
@@ -1222,6 +1219,9 @@ GO
 --     Creates or updates (including marking deleted) a FHIR resource
 --
 -- PARAMETERS
+--     @baseResourceSurrogateId
+--         * A bigint to which a value between [0, 80000) is added, forming a unique ResourceSurrogateId.
+--         * This value should be the current UTC datetime, truncated to millisecnd precision, with its 100ns ticks component bitshifted left by 3.
```
Comment (cunninghamjc) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/1.sql]: Typo: millisecnd

File: src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Persistence/ResourceHandlerTests.cs [hunk 3]
```diff
@@ -121,7 +121,7 @@ public async Task GivenAFhirMediator_WhenSavingAResource_ThenLastUpdatedShouldBe
 
                 resource = (await _mediator.UpsertResourceAsync(resource)).Resource;
 
-                Assert.Equal(instant, resource.LastUpdated);
+                Assert.Equal(new DateTimeOffset(DateTimeOffset.Now.Date.AddMilliseconds(6), TimeSpan.Zero), resource.LastUpdated);
```
Comment (cunninghamjc) [inline src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Persistence/ResourceHandlerTests.cs]: This could potentially fail around midnight?

Comment (johnstairs) [reply src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Persistence/ResourceHandlerTests.cs]: Yes, I should be snapshotting the `.Now` value. Good call!

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/ResourceSurrogateIdHelper.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+using System;
+using System.Diagnostics;
+using EnsureThat;
+using Microsoft.Health.Fhir.Core.Extensions;
+
+namespace Microsoft.Health.Fhir.SqlServer.Features.Storage
+{
+    /// <summary>
+    /// The resource surrogate ID is actually the modification datetime with a "uniquifier" added to it by the database.
```
Comment (brandonpollett) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Storage/ResourceSurrogateIdHelper.cs]: "modification datetime" -> "last modified datetime"?

**Pull request merged**